theano/compile/function.py

```diff
@@ -185,6 +186,22 @@ def opt_log1p(node):
 
 
     """
+    if isinstance(outputs, dict):
+        output_items = outputs.items()
+
+        output_items_sorted = sorted(output_items)
+
+        output_keys = []
+        outputs = []
+        for pair in output_items_sorted: 
```

Comment:
Why python 3 raise a different error is because you make the assert after the call to `sorted`. If you make the check for string before that call, all python version will raise an assertion error. You keep the creation of output_keys here.

Can you do this change to have the same behavior with python 2 and 3?